lib/puppet/pops/types/types.rb

```diff
@@ -666,12 +677,21 @@ def _assignable?(o, guard)
 # @api public
 #
 class PScalarType < PAnyType
+  KNOWN_SCALARS = [String, Integer, Float, TrueClass, FalseClass, Regexp, Time::Timestamp, Time::Timespan, SemanticPuppet::Version, SemanticPuppet::VersionRange]
+  KNOWN_NON_SCALARS = [Array, Hash, PAnyType, NilClass]
+
   def self.register_ptype(loader, ir)
     create_ptype(loader, ir, 'AnyType')
   end
 
   def instance?(o, guard = nil)
-    assignable?(TypeCalculator.infer(o), guard)
+    if KNOWN_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      true
+    elsif KNOWN_NON_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      false
```

Comment:
In my experience, a case/when is generally slower.  And Ruby will do a linear scan here, testing each expression with ===. There is no key to use for lookup.